File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/API_Sample_Code/advanced/Ruby/perf_counters/perf_aggr_latency.rb
#===============================================================#
#                                                               #
# ID                                                            #
#                                                               #
# perf_aggr_latency.rb                                          #
#                                                               #
# Copyright (c) 2011 NetApp, Inc. All rights reserved.          #
# Specifications subject to change without notice.              #
#                                                               #
# We do not have API to get read_latency, write_latency, and    #
# average_latency at aggregate level. This sample code          #
# demonstrates a method to get this data by reading the         #
# required latency of all volumes for the given aggregate.      #
# Use this information to generate latency at aggregate level.  #
#                                                               #
# This Sample code is supported from DataFabric Manager 3.7.1   #
# onwards.                                                      #
# However few of the functionalities of the sample code may     #
# work on older versions of DataFabric Manager.                 #
#===============================================================#

$:.unshift '../../../../../../../lib/ruby/NetApp'
require 'NaServer'


def print_help() 
	print ("\nCommand:\n")
	print ("perf_aggr_latency.rb <dfm> <user> <password> <aggr-name>\n")
	print ("<dfm>       -- DFM Server name\n")
	print ("<user>      -- User name\n")
	print ("<password>  -- Password\n")
	print ("<aggr-name> -- Name of the aggregate in format storage:aggrName\n")
	print ("-" * 80 + "\n")
	print ("This sample code provides information on read latency, write latency\n")
	print ("and average latency of an aggregate \n")
	print ("This data can be used to charts to represent data in graphical format.\n")
	print ("To generate the graph, redirect output of this sample code to\n")
	print ("an Excel sheet.\n")
	exit 
end
			

def print_output() 
	i = 0
	samples = $time1.length
	print ("Time\t\t\t\tRead Latency\tWrite Latency\tAverage Latency\n")	
	if (samples > 0) 	
		while(i < samples)
			print($time1[i])
			printf("\t%.4f\t\t%.4f\t\t%.4f\n",($read_latency[i].to_f/samples), ($write_latency[i].to_f/samples), ($avg_latency[i].to_f/samples))
			i = i + 1
		end
	end
end


def invoke_perf_zapi() 
	dfm = ARGV[0]
	user = ARGV[1]
	pw  = ARGV[2]
	aggr_name = ARGV[3]	
	# Initialize server context
	server_ctx = NaServer.new(dfm, 1, 0)
	server_ctx.set_admin_user(user, pw)
	server_ctx.set_server_type("DFM")
	# Create API request
	perf_in = NaElement.new("perf-get-counter-data")
	perf_in.child_add_string("duration", 6000)
	perf_in.child_add_string("number-samples", 50)
	perf_in.child_add_string("time-consolidation-method", "average")
	instance_info = NaElement.new("instance-counter-info")
	instance_info.child_add_string("object-name-or-id", aggr_name)
	counter_info = NaElement.new("counter-info")
	perf_obj_ctr1 = NaElement.new("perf-object-counter")
	perf_obj_ctr1.child_add_string("object-type", "volume")
	perf_obj_ctr1.child_add_string("counter-name", "read_latency")
	perf_obj_ctr2 = NaElement.new("perf-object-counter")
	perf_obj_ctr2.child_add_string("object-type", "volume")
	perf_obj_ctr2.child_add_string("counter-name", "write_latency")
	perf_obj_ctr3 = NaElement.new("perf-object-counter")
	perf_obj_ctr3.child_add_string("object-type", "volume")
	perf_obj_ctr3.child_add_string("counter-name", "avg_latency")
	counter_info.child_add(perf_obj_ctr1)
	counter_info.child_add(perf_obj_ctr2)
	counter_info.child_add(perf_obj_ctr3)
	instance_info.child_add(counter_info)
	perf_in.child_add(instance_info)
	perf_out = server_ctx.invoke_elem(perf_in)	
	if(perf_out.results_status() == "failed") 
		print(perf_out.results_reason() + "\n")
		exit
	end	
	return perf_out
end


def extract_perf_counter_data(perf_out) 
	gen_time_arr = 1
	time_i = 0
	read_i = 0
	write_i = 0
	avg_i = 0
	instance = perf_out.child_get("perf-instances")
	instances = instance.children_get()
	instances.each do |rec|
		vol_name = rec.child_get_string("instance-name")
		counters = rec.child_get("counters")
		perf_cnt_data = counters.children_get()		
		perf_cnt_data.each do |rec1|
			read_i = 0
			write_i = 0
			avg_i = 0
			counter_name = rec1.child_get_string("counter-name")
			counter_str = rec1.child_get_string("counter-data")
			counter_arr = counter_str.split(',')
			if(counter_name == "read_latency") 				
				counter_arr.each do |time_val|
					time_val_arr = time_val.split(':')					
					if(gen_time_arr) 
						$time1[time_i] = Time.at(time_val_arr[0].to_i)
						time_i = time_i + 1
					end
					if($read_latency[read_i]) 
						$read_latency[read_i] = $read_latency[read_i] + time_val_arr[1].to_f
					else 
						$read_latency[read_i] = time_val_arr[1].to_f
					end					
					read_i = read_i + 1
				end				
				gen_time_arr = nil	
				
			elsif(counter_name == "write_latency") 				
				counter_arr.each do |time_val|
					time_val_arr = time_val.split(':')					
					if(gen_time_arr) 
						$time1[time_i] = Time.at(time_val_arr[0].to_i)
						time_i = time_i + 1
					end					
					if($write_latency[write_i]) 
						$write_latency[write_i] = $write_latency[write_i] + time_val_arr[1].to_f					
					else 
						$write_latency[write_i] = time_val_arr[1].to_f
					end					
					write_i = write_i + 1
				end				
				gen_time_arr = nil
			
			elsif(counter_name == "avg_latency") 				
				counter_arr.each do |time_val|
					time_val_arr = time_val.split(':')					
					if(gen_time_arr) 
						$time1[time_i] = Time.at(time_val_arr[0].to_f)
						time_i = time_i + 1
					end					
					if ($avg_latency[avg_i]) 
						$avg_latency[avg_i] = $avg_latency[avg_i] + time_val_arr[1].to_f			
					else 
						$avg_latency[avg_i] = time_val_arr[1].to_f
					end					
					avg_i = avg_i + 1
				end				
				gen_time_arr = nil
			end
		end
	end
end


def main() 
	# check for valid number of parameters
	if ($args != 4)	
		print_help()
	end	
	perf_out = invoke_perf_zapi()
	extract_perf_counter_data(perf_out)
	print_output()
	exit 
end

	
# Command line arguments
$args = ARGV.length
$time1 = []
$read_latency = []
$write_latency = []
$avg_latency = []
#Invoke routine
main()

#=========================== POD ============================#

'=head1 NAME

  perf_aggr_latency.rb - This sample code provides information on read latency,
  write latency and average latency of an aggregate.

=head1 SYNOPSIS

  perf_aggr_latency.rb <dfm> <user> <password> <aggr-name>

=head1 ARGUMENTS

  <dfm>
  DFM Server name.

  <user>
  User name.

  <password>
  Password.

  <aggr-name>
  Name of the aggregate in format storage:aggrName.

=head1 SEE ALSO

  NaElement.rb, NaServer.rb

=head1 COPYRIGHT

 Copyright (c) 2009 NetApp, Inc. All rights reserved.
 Specifications subject to change without notice.

=cut
'

File: netapp-manageability-sdk-9.3/src/sample/DataFabric_Manager/API_Sample_Code/advanced/Java/perf_counters/perfAggrLatency.java
/**
 * $Id:$
 *
 * perfAggrLatency.java
 *
 * Copyright (c) 2009 NetApp, Inc. All rights reserved.
 * Specifications subject to change without notice.
 *
 *
 * This Sample code is supported from DataFabric Manager 3.7.1
 * onwards.
 * However few of the functionalities of the sample code may
 * work on older versions of DataFabric Manager.
 */

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.text.DecimalFormat;

import netapp.manage.NaAPIFailedException;
import netapp.manage.NaElement;
import netapp.manage.NaServer;

public class perfAggrLatency {
    public static void print_help() {
        System.out.println("\nCommand:");
        System.out.println("java perfAggrLatency <dfm> <user> <password>"
                + " <aggr_name>");
        System.out.println("<dfm>       -- DFM Server name");
        System.out.println("<user>      -- User name");
        System.out.println("<password>  -- Password");
        System.out.println("<aggr-name> -- Name of the aggregate in format"
                + " storage:aggrName");
        System.out.println("-----------------------------------------------"
                + "---------------------\n");
        System.out.println("This sample code provides information on read"
                + " latency, write latency");
        System.out.println("and average latency of an aggregate.");
        System.out.println("This data can be used to charts to represent"
                + " data in graphical format.");
        System.out.println("To generate the graph, redirect output of this"
                + " sample code to");
        System.out.println("an Excel sheet.");

        System.exit(1);
    }

    /**
     * getAggrLatency : Calculates the aggregate latency for all the volumes in
     * the aggregate specified.
     * 
     * @param server
     *            DFM server that is contacted to retrieve the information
     * @param aggr
     *            Parent aggregate, latency of volumes under this aggregate are
     *            calculated
     * @return This method outputs the values by itself, no return values
     */

    public static void getAggrLatency(NaServer server, String aggr) {
        NaServer connectedServer = server;
        String targetAggr = aggr;
        NaElement perfApiInput, instanceInfo, counterInfo;
        NaElement perfObjectCounter1, perfObjectCounter2, perfObjectCounter3;
        NaElement perfApiOutput;
        String localTime;
        int sample = 50;
        /*
         * Construct the API to be called here. (perf-get-counter-data)
         * Specified : the duration for api to execute, number of samples and
         * the method of collecting data.
         */

        perfApiInput = new NaElement("perf-get-counter-data");
        perfApiInput.addNewChild("duration", "6000");
        perfApiInput.addNewChild("number-samples", Integer.toString(sample));
        perfApiInput.addNewChild("time-consolidation-method", "average");

        instanceInfo = new NaElement("instance-counter-info");
        instanceInfo.addNewChild("object-name-or-id", targetAggr);

        counterInfo = new NaElement("counter-info");

        perfObjectCounter1 = new NaElement("perf-object-counter");
        perfObjectCounter1.addNewChild("object-type", "volume");
        perfObjectCounter1.addNewChild("counter-name", "read_latency");

        perfObjectCounter2 = new NaElement("perf-object-counter");
        perfObjectCounter2.addNewChild("object-type", "volume");
        perfObjectCounter2.addNewChild("counter-name", "write_latency");

        perfObjectCounter3 = new NaElement("perf-object-counter");
        perfObjectCounter3.addNewChild("object-type", "volume");
        perfObjectCounter3.addNewChild("counter-name", "avg_latency");

        counterInfo.addChildElem(perfObjectCounter1);
        counterInfo.addChildElem(perfObjectCounter2);
        counterInfo.addChildElem(perfObjectCounter3);
        instanceInfo.addChildElem(counterInfo);
        perfApiInput.addChildElem(instanceInfo);

        try {
            perfApiOutput = connectedServer.invokeElem(perfApiInput);

            NaElement iterValue;
            boolean generateTimeArray = true;
            ArrayList timeArray = new ArrayList();
            double[] readDataArray = new double[sample];
            double[] writeDataArray = new double[sample];
            double[] avgDataArray = new double[sample];
            int timeIter, readIter, writeIter, avgIter;

            List instances = perfApiOutput.getChildByName("perf-instances")
                    .getChildren();

            Iterator instanceIter = instances.iterator();

            /*
             * Populate the counter data for aggregate latency.
             */

            while (instanceIter.hasNext()) {

                String counterName;
                readIter = 0;
                writeIter = 0;
                avgIter = 0;
                iterValue = (NaElement) instanceIter.next();

                if (iterValue.getChildByName("counters") != null) {
                    List counters1 = iterValue.getChildByName("counters")
                            .getChildren();
                    Iterator counterIter1 = counters1.iterator();

                    while (counterIter1.hasNext()) {

                        NaElement counter1 = (NaElement) counterIter1.next();
                        String counterData = counter1
                                .getChildContent("counter-data");
                        counterName = counter1.getChildContent("counter-name");

                        String[] counterArray;
                        int iter = 0;
                        counterArray = counterData.split(",");
                        if (counterName.equals("read_latency")) {
                            for (iter = 0; iter < counterArray.length; iter++) {
                                String[] counterDataArray;
                                counterDataArray = counterArray[iter]
                                        .split(":");

                                if (generateTimeArray == true) {
                                    localTime = new SimpleDateFormat(
                                            "dd/MM/yyyy HH:mm:ss")
                                            .format(new Date(
                                                    Long.parseLong(counterDataArray[0]) * 1000));
                                    timeArray.add(localTime);
                                }
                                readDataArray[readIter] += Double
                                        .parseDouble(counterDataArray[1]);
                                readIter++;
                            }
                            generateTimeArray = false;
                        } else if (counterName.equals("write_latency")) {
                            for (iter = 0; iter < counterArray.length; iter++) {
                                String[] counterDataArray;
                                counterDataArray = counterArray[iter]
                                        .split(":");

                                if (generateTimeArray == true) {
                                    localTime = new SimpleDateFormat(
                                            "dd/MM/yyyy HH:mm:ss")
                                            .format(new Date(
                                                    Long.parseLong(counterDataArray[0]) * 1000));
                                    timeArray.add(localTime);
                                }
                                writeDataArray[writeIter] += Double
                                        .parseDouble(counterDataArray[1]);
                                writeIter++;
                            }
                            generateTimeArray = false;
                        } else if (counterName.equals("avg_latency")) {
                            for (iter = 0; iter < counterArray.length; iter++) {
                                String[] counterDataArray;
                                counterDataArray = counterArray[iter]
                                        .split(":");

                                if (generateTimeArray == true) {
                                    localTime = new SimpleDateFormat(
                                            "dd/MM/yyyy HH:mm:ss")
                                            .format(new Date(
                                                    Long.parseLong(counterDataArray[0]) * 1000));
                                    timeArray.add(localTime);
                                }
                                avgDataArray[avgIter] += Double
                                        .parseDouble(counterDataArray[1]);
                                avgIter++;
                            }
                            generateTimeArray = false;
                        }
                    }
                }
            }
            /*
             * Print data for aggregate latencies.
             */

            System.out.println("Time\tRead Latency\tWrite Latency\tAverage"
                    + " Latency");
            int total = timeArray.size();
            DecimalFormat f = new DecimalFormat("#####.####");

            for (timeIter = 0; timeIter < total; timeIter++) {
                System.out.println(timeArray.get(timeIter) + "\t"
                        + f.format(readDataArray[timeIter] / total) + "\t"
                        + f.format(writeDataArray[timeIter] / total) + "\t"
                        + f.format(avgDataArray[timeIter] / total));
            }

        } catch (NaAPIFailedException exception) {
            exception.printStackTrace();
            System.exit(1);
        } catch (Exception exception) {
            exception.printStackTrace();
            System.exit(1);
        }
    }

    /**
     * @param args
     */
    public static void main(String[] args) {

        NaServer server;

        if (args.length < 4) {
            print_help();
        }

        String aggr = new String(args[3]);

        try {
            server = new NaServer(args[0], 1, 0);
            server.setStyle(NaServer.STYLE_LOGIN_PASSWORD);
            server.setAdminUser(args[1], args[2]);
            server.setServerType(NaServer.SERVER_TYPE_DFM);
            server.setPort(8088);

            getAggrLatency(server, aggr);

        } catch (Exception exception) {
            exception.printStackTrace();
            System.exit(1);
        }
    }

}
